src/components/Tag.ts

```diff
@@ -0,0 +1,151 @@
+        };
+        return createElement("div",
+            {
+                className: classNames(
+                    "widget-tag",
+                    `widget-tag-${this.props.tagStyle}`,
+                    this.props.className
+                ),
+                style: this.props.style
+            },
+            createElement(TagsInput, {
+                addOnBlur: true,
+                addOnPaste: true,
+                disabled: this.props.readOnly,
+                inputProps,
+                inputValue: this.state.tag,
+                onChangeInput: this.handleChangeInput,
+                maxTags: this.props.tagLimit === 0 ? 1000 : this.props.tagLimit,
+                onChange: this.handleChange,
+                renderInput: this.props.enableSuggestions ? this.autosuggest : undefined,
+                value: this.state.tags
+            }),
+            createElement(Alert, {
+                bootstrapStyle: "danger",
+                className: "widget-tag-alert",
+                message: this.state.alertMessage
+            })
+        );
+    }
+
+    private handleChangeInput(tag: string) {
+        const { tagLimit, tagLimitMessage, showError } = this.props;
+        if (tagLimit === 0) {
+            this.setState({ tag });
+        } else if (this.state.tags.length >= tagLimit) {
+            showError(tagLimitMessage.replace("{limit}", `${tagLimit}`));
+            this.setState({ tag: "" });
+        } else {
+            this.setState({ tag });
+        }
+    }
+
+    private getSuggestions(): Suggestion[] {
+        const suggestions: Suggestion[] = this.props.tags.map(tag => ({
+            name: tag,
+            newValue: "",
+            suggestionValue: "",
+            value: ""
+        }));
+
+        return suggestions;
+    }
+
+    private autosuggest() {
+        return createElement(AutoComplete, {
+            addTag: (tag: string) => this.addTag(tag),
+            fetchSuggestions: this.props.lazyLoadTags,
+            inputPlaceholder: this.props.inputPlaceholder,
+            lazyLoad: this.props.lazyLoad,
+            suggestions: this.getSuggestions()
+        });
+    }
+
+    private addTag(tag: string) {
+        const { enableCreate, tags, tagLimit, tagLimitMessage, createTag } = this.props;
+        if (tagLimit === 0 || tagLimit > this.state.tags.length) {
+            if (enableCreate && createTag && tags.indexOf(tag) === -1 && this.state.tags.indexOf(tag) === -1) {
+                createTag(tag);
+            }
+            this.state.tags.push(tag);
+            this.setState({ tags: this.state.tags });
+        } else {
+            this.props.showError(tagLimitMessage.replace("{limit}", `${tagLimit}`));
+        }
+    }
+
+    private handleChange(tags: string[], changed: string[]) {
+        if (this.state.tags.length > tags.length) {
+            this.setState({ tags });
+        } else {
+            this.addTag(changed[0]);
```

Comment:
can there be more than one? or worse... none?

Reply:
Done